### PR TITLE
Add a switchable protocol scheme in order to use HTTP or HTTPS

### PR DESCRIPTION
**Pull request by kcm**
Based on https://github.com/elastic/ansible-elasticsearch/pull/302

### CODE REVIEW

**Comment by Crazybus** [comment]
jenkins test this please

**Comment by Crazybus** [comment]
jenkins test this please

**Comment by Crazybus** [comment]
Hah. So the license was expired, but now it is still failing on the license install but for a real reason.


```
Cannot install a [GOLD] license unless TLS is configured or security is disabled
```

To reproduce this I'm running:
```
make verify VERSION=6.x PATTERN=xpack-upgrade-ubuntu-1604
```

And setting `no_log: False` in `tasks/xpack/security/elasticsearch-xpack-activation.yml`

Full output:

```
      TASK [elasticsearch : Activate ES license (with security authentication)] ******
       fatal: [localhost]: FAILED! => {"changed": false, "content": "{\"error\":{\"root_cause\":[{\"type\":\"illegal_state_exception\",\"reason\":\"Cannot install a [GOLD] license unless TLS is configured or security is disabled\"}],\"type\":\"illegal_state_exception\",\"reason\":\"Cannot install a [GOLD] license unless TLS is configured or security is disabled\"},\"status\":500}", "content_length": "291", "content_type": "application/json; charset=UTF-8", "failed_when_result": true, "json": {"error": {"reason": "Cannot install a [GOLD] license unless TLS is configured or security is disabled", "root_cause": [{"reason": "Cannot install a [GOLD] license unless TLS is configured or security is disabled", "type": "illegal_state_exception"}], "type": "illegal_state_exception"}, "status": 500}, "msg": "Status code was not [200]: HTTP Error 500: Internal Server Error", "redirected": false, "status": 500, "url": "http://localhost:9200/_xpack/license?acknowledge=true"}
       	to retry, use: --limit @/tmp/kitchen/xpack-upgrade.retry
```

However this isn't something you changed. It looks like the tests were always using a trial license which allows you to enable security without using TLS. I'm going to fix this up since this is broken due to the test license changes.

**Comment by visokoo** [comment]
@Crazybus Any movement on this from your end on fixing the tests? I'm seeing a need for this as well since we're also enabling https for api calls on our cluster.

File: tasks/xpack/security/elasticsearch-xpack-activation.yml [hunk 1]
```diff
@@ -18,7 +18,7 @@
 - name: Activate ES license (with security authentication)
```
Comment (Crazybus) [inline tasks/xpack/security/elasticsearch-xpack-activation.yml]: The license was fixed a while back however this task is still failing. I believe it is because there isn't a valid certificate. This task will need `validate_certs: no` to make sure that the license can be installed. I saw that you added similiar logic to the actual serverspec tests so that will need to be duplicated here.

**Comment by itsmed** [comment]
I'm taking at look at the tests as well, and I think I'm running into issues with OpenSSL.

**Comment by itsmed** [comment]
jenkins test this please

**Comment by ahulab** [comment]
hello, has this been forgotten about?

This is something that has bitten me a number of times now, and it looks like there is a lot of interest from the community to get this working. I see a bunch of open issues and PRs about this, and so I'm commenting here as to not create new items. 
- https://github.com/elastic/ansible-elasticsearch/issues/500
- https://github.com/elastic/ansible-elasticsearch/issues/579
- https://github.com/elastic/ansible-elasticsearch/issues/278
- https://github.com/elastic/ansible-elasticsearch/pull/442
- https://github.com/elastic/ansible-elasticsearch/pull/302

API interaction via HTTPS is now especially useful since 6.8.x allows users to use security features with the basic license. My current workflow is to configure the cluster first without ssl enabled, so that all API interactions work, and then afterwards I run my playbook again and enable SSL. This has obvious shortcomings. 

I am more than happy to work with the maintainers to get this fixed, if someone could provide some guidance that would be great. I can open a separate PR, or address issues in a current but abandoned PR, whatever works.

**Comment by itsmed** [comment]
Thanks for the ping, I'll get back to this this week.

**Comment by itsmed** [comment]
jenkins test this

**Comment by itsmed** [comment]
@Crazybus @jmlrt I've been unable to get this to pass due to a bug that seems to interpret the license.json as malformed json, if the es devs I've pulled in helped to diagnose correctly. However, I've added verbosity and verified that the license is being sent correctly, and has the same format as using `curl -d @license.json ...` syntax. Furthermore I've been unable to reproduce when using a docker container for es ( rather than the kitchen/ansible tests ). Can we put some time together to sync and take a look at this together?

File: tasks/xpack/security/elasticsearch-xpack-activation.yml [hunk 2]
```diff
@@ -14,6 +15,24 @@
...
+
+- name: Activate ES license (with security authentication)
+  uri:
+    method: PUT
+    url: "{{es_api_scheme}}://{{es_api_host}}:{{es_api_port}}/_license?acknowledge=true"
+    user: "{{es_api_basic_auth_username  | default(omit)}}"
+    password: "{{es_api_basic_auth_password | default(omit)}}"
+    body_format: json
+    body: "{{ es_xpack_license }}"
+    return_content: yes
```
Comment (Crazybus) [inline tasks/xpack/security/elasticsearch-xpack-activation.yml]: You are going to need to either point to the actual certs for this request or disable the validation with: `validate_certs: no`.

**Comment by Crazybus** [comment]
> @Crazybus @jmlrt I've been unable to get this to pass due to a bug that seems to interpret the license.json as malformed json, if the es devs I've pulled in helped to diagnose correctly. However, I've added verbosity and verified that the license is being sent correctly, and has the same format as using curl -d @license.json ... syntax. Furthermore I've been unable to reproduce when using a docker container for es ( rather than the kitchen/ansible tests ). Can we put some time together to sync and take a look at this together?

Can you post the exact curl command that you are using to test with? I think it could just be a cert validation issue (I left a comment on the PR where I think it is). But lets absolutely setup a time to sync on this and figure it out. It looks like you are really really close! And thank you so much for working on this.

**Comment by Crazybus** [comment]
I also just saw that the master build is failing with this error:

```
05:40:20        TASK [elasticsearch : Activate ES license (with security authentication)] ******
05:40:21        fatal: [localhost]: FAILED! => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", "changed": false}
```

If it works with curl then it seems unlikely to be related. But just thought I would mention it here since it seems like too much of a coincidence.

**Comment by Crazybus** [comment]
The reason why the master build is failing: https://github.com/elastic/ansible-elasticsearch/pull/588#issuecomment-509771246 is actually a reason why your testing *should* be working. So I don't think it was related.

**Comment by itsmed** [comment]
I've run the tests using `$ make verify VERSION=7.x PATTERN=xpack-upgrade-centos-7`
and then `exec`'d into the container
```
sh-4.2# cat /etc/*-release
CentOS Linux release 7.6.1810 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.6.1810 (Core)
CentOS Linux release 7.6.1810 (Core)
```
using `curl` from within the container after the tests failed locally
```
sh-4.2# curl -XPUT -u elastic:elasticChanged -H "Content-Type: application/json" -d @license.json localhost:9200/_license?acknowledge=true
{"acknowledged":true,"license_status":"valid"}sh-4.2#
```
FWIW this _is_ using a basic license, I see that you updated the test license so we're now using the same one. I should have pinged when I saw the tests were using a different license, but I didn't think that would be the problem 🤦‍♂ 

I'm rather sure that I've gotten the tasks portion of this PR sorted out, but next will be updating the test suite to accommodate for the changes in the API in 7.x .  I have to run some errands now, but I'll swing back to this later this evening.

**Comment by Crazybus** [comment]
In the configuration for this test you have the api scheme set to `https`:

```
es_api_scheme: "https"
```

But the curl command you showed is using `http`:

```
curl -XPUT -u elastic:elasticChanged -H "Content-Type: application/json" -d @license.json localhost:9200/_license?acknowledge=true
{"acknowledged":true,"license_status":"valid"}
```

I think the cleanest way to solve this is to actually have ansible test whether or not it can access Elasticsearch with the configured protocol, if not it can then try to fallback to `http` to make sure that it can actually install the license to allow `https`

**Comment by itsmed** [comment]
Thanks @Crazybus I hadn't pushed the latest changes, but I actually removed the`es_api_scheme` references in the test. I think the problem may be coming from the `net/http` module if I'm reading this correctly
```

       An error occurred while loading /tmp/verifier/suites/serverspec/default_spec.rb.
       Failure/Error: req.use_ssl = true
       NoMethodError:
         undefined method `use_ssl=' for #<Net::HTTP::Get GET>

       # /tmp/verifier/suites/serverspec/spec_helper.rb:23:in `curl_json'
       # /tmp/verifier/suites/serverspec/shared_spec.rb:55:in `block (2 levels) in <top (required)>'
       # /tmp/verifier/suites/serverspec/shared_spec.rb:46:in `block in <top (required)>'
       # /tmp/verifier/suites/serverspec/default_spec.rb:7:in `block in <top (required)>'
       # /tmp/verifier/suites/serverspec/default_spec.rb:6:in `<top (required)>'
```

but I was reading old docs, it seems. The [newer ones](https://docs.ruby-lang.org/en/2.0.0/Net/HTTP.html#class-Net::HTTP-label-HTTPS) mention that you don't need to require `net/https`.

**Comment by Crazybus** [comment]
> @Crazybus @jmlrt I've been unable to get this to pass due to a bug that seems to interpret the license.json as malformed json, if the es devs I've pulled in helped to diagnose correctly. However, I've added verbosity and verified that the license is being sent correctly, and has the same format as using curl -d @license.json ... syntax. Furthermore I've been unable to reproduce when using a docker container for es ( rather than the kitchen/ansible tests ). Can we put some time together to sync and take a look at this together?

Your original ping ☝️ made it sound like you couldn't get the license installed anymore. Is that part working now but only the test isn't failing? 

> Thanks @Crazybus I hadn't pushed the latest changes

Could you push the latest changes? Then I can play around with it to see what might be causing the issue.

**Comment by itsmed** [comment]
FWIW I am looking into why the `xpack-upgrade` test is failing for 6.x
```
  - name: show the current es_api_basic_auth_username
    debug:
      msg: "es_api_basic_auth_username: {{es_api_basic_auth_username}} ; native users: {{native_users[es_api_basic_auth_username]}}"
```

```
       TASK [elasticsearch : show the current es_api_basic_auth_username] *************
       [DEPRECATION WARNING]: evaluating manage_native_realm as a bare variable, this
       behaviour will go away and you might need to add |bool to the expression in the
        future. Also see CONDITIONAL_BARE_VARS configuration toggle.. This feature
       will be removed in version 2.12. Deprecation warnings can be disabled by
       setting deprecation_warnings=False in ansible.cfg.
       ok: [localhost] => {
           "msg": null
       }
```

🤔

**Comment by itsmed** [comment]
Alright, not sure what is going on there. I've verified from within the container the expected users exists, and that I can change the password. I've tried hardcoding the username into the api call and ran the make target again, and moved past this task
```
  - name: Update API User Password
    uri:
      url: "http://{{es_api_host}}:{{es_api_port}}/_xpack/security/user/elastic/_password"
      method: POST
      body_format: json
      body: "{ \"password\":\"{{native_users[es_api_basic_auth_username].password}}\" }"
      status_code: 200
      user: "{{es_api_basic_auth_username}}"
      password: "{{es_api_basic_auth_password}}"
      force_basic_auth: yes
    when: change_api_password
```
So obviously something I'm not seeing changed when adding the blocks.

**Comment by itsmed** [comment]
@Crazybus I've pushed what I have. As is, the `Update API User Password` task on 6.x fails with
```
"error": {
                   "reason": "Validation Failed: 1: user must exist in order to change password;",
                   "root_cause": [
                       {
                           "reason": "Validation Failed: 1: user must exist in order to change password;",
                           "type": "validation_exception"
                       }
                   ],
                   "type": "validation_exception"
               },
               "status": 400
           },
           "msg": "Status code was 400 and not [200]: HTTP Error 400: Bad Request"
```

the version of `tasks/xpack/security/elasticsearch-security-native.yml` in master ends with 
```
       An error occurred while loading /tmp/verifier/suites/serverspec/default_spec.rb.
       Failure/Error: req.use_ssl = true
       NoMethodError:
         undefined method `use_ssl=' for #<Net::HTTP::Get GET>

       # /tmp/verifier/suites/serverspec/spec_helper.rb:23:in `curl_json'
       # /tmp/verifier/suites/serverspec/shared_spec.rb:55:in `block (2 levels) in <top (required)>'
       # /tmp/verifier/suites/serverspec/shared_spec.rb:46:in `block in <top (required)>'
       # /tmp/verifier/suites/serverspec/default_spec.rb:7:in `block in <top (required)>'
       # /tmp/verifier/suites/serverspec/default_spec.rb:6:in `<top (required)>'
       WARNING: Shared example group 'shared::init' has been previously defined at:
         /tmp/verifier/suites/serverspec/shared_spec.rb:40
       ...and you are now defining it at:
         /tmp/verifier/suites/serverspec/shared_spec.rb:40
       The new definition will overwrite the original one.
       WARNING: Shared example group 'xpack_upgrade::init' has been previously defined at:
         /tmp/verifier/suites/serverspec/xpack_upgrade_spec.rb:5
       ...and you are now defining it at:
         /tmp/verifier/suites/serverspec/xpack_upgrade_spec.rb:5
       The new definition will overwrite the original one.

       Finished in 0.00127 seconds (files took 0.61664 seconds to load)
       0 examples, 0 failures, 1 error occurred outside of examples
```

Would you mind taking a look again?

**Comment by Crazybus** [comment]
When running it locally this is the error I'm getting:

```
TASK [elasticsearch : Update API User Password] ********************************
       fatal: [localhost]: FAILED! => {"changed": false, "content": "{\"error\":\"no handler found for uri [/_xpack/security/user/{[es_api_basic_auth_username}}/password] and method
 [POST]\"}", "content_length": "118", "content_type": "application/json; charset=UTF-8", "json": {"error": "no handler found for uri [/_xpack/security/user/{[es_api_basic_auth_usern
ame}}/password] and method [POST]"}, "msg": "Status code was 400 and not [200]: HTTP Error 400: Bad Request", "redirected": false, "status": 400, "url": "http://localhost:9200/_xpac
k/security/user/{[es_api_basic_auth_username}}/password"}
```
I think part of the problem is you seem to be looking at the wrong block. For 6.x it's actually running this task right?

https://github.com/elastic/ansible-elasticsearch/blob/efe8f0e91345b589b0505b74903c6cdfab0ccd81/tasks/xpack/security/elasticsearch-security-native.yml#L258-L268

So I think the first piece of the puzzle is to fix up `/password` to point to `/_password`. 

What is the reason for needing separate blocks for 6.x and 7.x? Is it only for adding removing the /xpack in the url? If so we could do something similar with how this was handled for some directory changes when xpack was opened up:

https://github.com/elastic/ansible-elasticsearch/blob/efe8f0e91345b589b0505b74903c6cdfab0ccd81/tasks/compatibility-variables.yml#L31

**Comment by Crazybus** [comment]
The next issue is that the path seems to contain the jinja2 ansible logic:

```
HTTP Error 400: Bad Request", "redirected": false, "status": 400, "url": "http://localhost:9200/_xpac
k/security/user/{[es_api_basic_auth_username}}/password
```
The path should be `http://localhost:9200/_xpack/security/user/elastic/password`

The issue is a small typo here:
```
     url: "http://{{es_api_host}}:{{es_api_port}}/_xpack/security/user/{[es_api_basic_auth_username}}/password" 
```

`{[es_api_basic_auth_username}}` should be `{{es_api_basic_auth_username}}`. A square bracket stuck in there which caused ansible to output the actual template instead of the rendered value.

**Comment by itsmed** [comment]
Cheers for that. I need to get linting set up in the new editor.
So after simplifying, and some typo fixes and removing the debugging code, here it is, the ruby `use_ssl` error I was talking about.

```
$ make verify VERSION=7.x PATTERN=xpack-upgrade-centos-7

       Could not find gem 'rspec-retry' in any of the gem sources listed in your
       Gemfile.
       Don't run Bundler as root. Bundler can ask for sudo if it is needed, and
       installing your bundle as root will break this application for all non-root
       users on this machine.
       Fetching gem metadata from https://rubygems.org/...........
       Resolving dependencies...
       Using bundler 2.0.2
       Using rspec-support 3.8.2
       Using rspec-core 3.8.2
       Fetching rspec-retry 0.6.1
       Installing rspec-retry 0.6.1
       Bundle complete! 1 Gemfile dependency, 4 gems now installed.
       Use `bundle info [gemname]` to see where a bundled gem is installed.
-----> Installing Serverspec..
       Fetching diff-lcs-1.3.gem
       Fetching rspec-3.8.0.gem
       Fetching rspec-expectations-3.8.4.gem
       Fetching rspec-its-1.3.0.gem
       Fetching multi_json-1.13.1.gem
       Fetching net-ssh-5.2.0.gem
       Fetching net-scp-2.0.0.gem
       Fetching rspec-mocks-3.8.1.gem
       Fetching net-telnet-0.1.1.gem
       Fetching sfl-2.3.gem
       Fetching specinfra-2.79.0.gem
       Fetching serverspec-2.41.4.gem
-----> serverspec installed (version 2.41.4)
       /opt/chef/embedded/bin/ruby -I/tmp/verifier/suites/serverspec -I/tmp/verifier/gems/gems/rspec-support-3.8.2/lib:/tmp/verifier/gems/gems/rspec-core-3.8.2/lib /opt/chef/embedded/bin/rspec --pattern /tmp/verifier/suites/serverspec/\*\*/\*_spec.rb --color --format documentation --default-path /tmp/verifier/suites/serverspec

       An error occurred while loading /tmp/verifier/suites/serverspec/default_spec.rb.
       Failure/Error: req.use_ssl = true
       NoMethodError:
         undefined method `use_ssl=' for #<Net::HTTP::Get GET>

       # /tmp/verifier/suites/serverspec/spec_helper.rb:23:in `curl_json'
       # /tmp/verifier/suites/serverspec/shared_spec.rb:55:in `block (2 levels) in <top (required)>'
       # /tmp/verifier/suites/serverspec/shared_spec.rb:46:in `block in <top (required)>'
       # /tmp/verifier/suites/serverspec/default_spec.rb:7:in `block in <top (required)>'
       # /tmp/verifier/suites/serverspec/default_spec.rb:6:in `<top (required)>'
       WARNING: Shared example group 'shared::init' has been previously defined at:
         /tmp/verifier/suites/serverspec/shared_spec.rb:40
       ...and you are now defining it at:
         /tmp/verifier/suites/serverspec/shared_spec.rb:40
       The new definition will overwrite the original one.
       WARNING: Shared example group 'xpack_upgrade::init' has been previously defined at:
         /tmp/verifier/suites/serverspec/xpack_upgrade_spec.rb:5
       ...and you are now defining it at:
         /tmp/verifier/suites/serverspec/xpack_upgrade_spec.rb:5
       The new definition will overwrite the original one.

       Finished in 0.0003 seconds (files took 0.48334 seconds to load)
       0 examples, 0 failures, 1 error occurred outside of examples
```
and
```
$ make verify VERSION=6.x PATTERN=xpack-upgrade-centos-7

       An error occurred while loading /tmp/verifier/suites/serverspec/default_spec.rb.
       Failure/Error: req.use_ssl = true
       NoMethodError:
         undefined method `use_ssl=' for #<Net::HTTP::Get GET>

       # /tmp/verifier/suites/serverspec/spec_helper.rb:23:in `curl_json'
       # /tmp/verifier/suites/serverspec/shared_spec.rb:55:in `block (2 levels) in <top (required)>'
       # /tmp/verifier/suites/serverspec/shared_spec.rb:46:in `block in <top (required)>'
       # /tmp/verifier/suites/serverspec/default_spec.rb:7:in `block in <top (required)>'
       # /tmp/verifier/suites/serverspec/default_spec.rb:6:in `<top (required)>'
       WARNING: Shared example group 'shared::init' has been previously defined at:
         /tmp/verifier/suites/serverspec/shared_spec.rb:40
       ...and you are now defining it at:
         /tmp/verifier/suites/serverspec/shared_spec.rb:40
       The new definition will overwrite the original one.
       WARNING: Shared example group 'xpack_upgrade::init' has been previously defined at:
         /tmp/verifier/suites/serverspec/xpack_upgrade_spec.rb:5
       ...and you are now defining it at:
         /tmp/verifier/suites/serverspec/xpack_upgrade_spec.rb:5
       The new definition will overwrite the original one.

       Finished in 0.00029 seconds (files took 0.45504 seconds to load)
       0 examples, 0 failures, 1 error occurred outside of examples
```

I'll poke some more this weekend, now that we're back to this point.

**Comment by Crazybus** [comment]
I'm not sure why that even needs to be set, each of the calls starts with `http://` or `https://` so the library should be able to figure it out.

Removing it got the tests passing for me. Here is what I have locally. 
```
def curl_json(uri, username=nil, password=nil)
  uri = URI(uri)
  req =  Net::HTTP::Get.new(uri)
  if username && password
    req.basic_auth username, password
  end
  res = Net::HTTP.start(
    uri.hostname,
    uri.port,
    :verify_mode => OpenSSL::SSL::VERIFY_NONE
  ) {|http|
    http.request(req)
  }
  return JSON.parse(res.body)
end
```

```
       Finished in 3.34 seconds (files took 0.86895 seconds to load)
       60 examples, 0 failures
```

**Comment by itsmed** [comment]
@Crazybus @jmlrt CI is blue!

**Comment by jmlrt** [comment]
@itsmed That's awesome! Sorry I didn't followed it so much. Do you have any feature done for this PR and are you ready for some review?

**Comment by itsmed** [comment]
No worries @jmlrt when you guys have time, please take a look, and sorry for all the noise!

File: README.md [hunk 3]
```diff
@@ -223,6 +223,12 @@ An example of a three server deployment is shown below.  The first server holds
       node.data: true
       node.master: false
       bootstrap.memory_lock: false
+    es_scripts: false
```
Comment (Crazybus) [inline README.md]: This variable doesn't exist anywhere else in the playbook. Why did you add it here?

File: README.md [hunk 4]
```diff
@@ -223,6 +223,12 @@ An example of a three server deployment is shown below.  The first server holds
...
+    es_templates: false
```
Comment (Crazybus) [inline README.md]: This is already the default value. Why add it to the example as well?

File: README.md [hunk 5]
```diff
@@ -223,6 +223,12 @@ An example of a three server deployment is shown below.  The first server holds
...
+    es_version_lock: false
```
Comment (Crazybus) [inline README.md]: Another default

File: README.md [hunk 6]
```diff
@@ -223,6 +223,12 @@ An example of a three server deployment is shown below.  The first server holds
...
+    ansible_user: ansible
```
Comment (Crazybus) [inline README.md]: Same question for this one, except this will actually causes breakages for anyone not using the ansible user for ssh.

File: README.md [hunk 7]
```diff
@@ -223,6 +223,12 @@ An example of a three server deployment is shown below.  The first server holds
...
+    es_api_scheme: "http"
+    es_api_port: 9200
```
Comment (Crazybus) [inline README.md]: This one is also the default and can be removed.

File: tasks/elasticsearch-template.yml [hunk 8]
```diff
@@ -15,15 +15,30 @@
...
     force_basic_auth: yes
     body_format: json
     body: "{{ lookup('file', item) }}"
-  when: load_templates.changed and es_start_service
+  when: load_templates.changed and es_start_service and not es_enable_xpack or not es_xpack_features is defined or "security" not in es_xpack_features
+  with_fileglob:
+    - "{{ es_templates_fileglob | default('') }}"
+  run_once: True
+
+- name: Install templates with auth
```
Comment (Crazybus) [inline tasks/elasticsearch-template.yml]: I think this changed since you started working on this. The above task was refactored to work for auth and no auth by doing `user: "{{es_api_basic_auth_username | default(omit)}}"`

File: tasks/elasticsearch-template.yml [hunk 9]
```diff
@@ -15,15 +15,30 @@
...
+    url: "{{es_api_uri}}/_template/{{item | filename}}"
     method: PUT
     status_code: 200
     user: "{{es_api_basic_auth_username | default(omit)}}"
     password: "{{es_api_basic_auth_password | default(omit)}}"
     force_basic_auth: yes
     body_format: json
     body: "{{ lookup('file', item) }}"
-  when: load_templates.changed and es_start_service
+  when: load_templates.changed and es_start_service and not es_enable_xpack or not es_xpack_features is defined or "security" not in es_xpack_features
```
Comment (Crazybus) [inline tasks/elasticsearch-template.yml]: Did you mean to revert this? We recently simplified it in https://github.com/elastic/ansible-elasticsearch/commit/a677218fa85a5cbbc7e835a3ede8e5a6c7cee2bf#diff-8cdc0121d463a439bd46901832ee0c2f

File: tasks/elasticsearch-template.yml (hunk 8 above)
Comment (Crazybus) [reply tasks/elasticsearch-template.yml]: https://github.com/elastic/ansible-elasticsearch/commit/a677218fa85a5cbbc7e835a3ede8e5a6c7cee2bf#diff-8cdc0121d463a439bd46901832ee0c2f is where it was changed, this commit seems to revert it.

File: tasks/xpack/security/elasticsearch-security-native.yml [hunk 10]
```diff
@@ -16,12 +16,30 @@
   set_fact: manage_native_roles=true
```
Comment (Crazybus) [inline tasks/xpack/security/elasticsearch-security-native.yml]: Commenting here since I can't do it above. The vim swap file `tasks/xpack/security/.elasticsearch-security-native.yml.swp` is checked in.

File: tasks/xpack/security/elasticsearch-security-native.yml [hunk 11]
```diff
@@ -16,12 +16,30 @@
...
+    method: GET
+    user: "{{es_api_basic_auth_username}}"
+    password: "{{es_api_basic_auth_password}}"
+    force_basic_auth: yes
+    status_code: 200
+  register: user_list_response
+  when: manage_native_users
+  check_mode: no
+
+- name: List Native Users
```
Comment (Crazybus) [inline tasks/xpack/security/elasticsearch-security-native.yml]: This task is a duplicate from the one above.

File: tasks/xpack/security/elasticsearch-security-native.yml [hunk 12]
```diff
@@ -16,12 +16,30 @@
...
+  set_fact:
+    es_security_endpoint: "{{'_security' if es_version is version_compare('7.0', operator='>=') else '_xpack/security'}}"
+
 #If the node has just has security installed it maybe either stopped or started 1. if stopped, we need to start to load native realms 2. if started, we need to restart to load
 
 #List current users
 - name: List Native Users
   uri:
-    url: http://{{es_api_host}}:{{es_api_port}}/_xpack/security/user
+    url: "http://{{es_api_host}}:{{es_api_port}}/{{es_security_endpoint}}/user"
```
Comment (Crazybus) [inline tasks/xpack/security/elasticsearch-security-native.yml]: I'm would need to confirm this locally, but I believe that this being hardcoded to http means it will fail on the second run. We don't have the [idempotency_test](https://github.com/elastic/ansible-elasticsearch/blob/5b1d028bd2f81e343f7dc90c4f5b8b23ad8e80e5/.kitchen.yml#L130) enabled for this specific suite because it is supposed to test an upgrade path. So the tests pass because it first runs with http, then with https, though I fear it wil fail on the third fun when Elasticsearch is listening on https at this point

File: tasks/xpack/security/elasticsearch-security-native.yml [hunk 13]
```diff
@@ -49,9 +67,20 @@
   set_fact: change_api_password=true
   when: manage_native_users and es_api_basic_auth_username in native_users and native_users[es_api_basic_auth_username].password is defined
 
+- name: List Native Users
```
Comment (Crazybus) [inline tasks/xpack/security/elasticsearch-security-native.yml]: Could you give this a different task name? It took me a little while to figure out the difference between this task and the ones above since they are very similar.

File: tasks/xpack/security/elasticsearch-security-native.yml [hunk 14]
```diff
@@ -73,7 +102,7 @@
 #Delete all non required users NOT inc. reserved
 - name: Delete Native Users
   uri:
-    url: http://{{es_api_host}}:{{es_api_port}}/_xpack/security/user/{{item}}
+    url: "https://{{es_api_host}}:{{es_api_port}}/{{es_security_endpoint}}/user/{{item}}"
```
Comment (Crazybus) [inline tasks/xpack/security/elasticsearch-security-native.yml]: I don't think this one can be hardcoded as it will break for anybody using `es_api_scheme: "http"`.

File: tasks/xpack/security/elasticsearch-xpack-activation.yml [hunk 15]
```diff
@@ -15,6 +16,25 @@
     license_activated.status != 200 or
     license_activated.json.license_status is not defined or
     license_activated.json.license_status != 'valid'
+  when: es_version is version('6.9', operator='<=')
+
+- name: Activate ES license (with security authentication)
```
Comment (Crazybus) [inline tasks/xpack/security/elasticsearch-xpack-activation.yml]: This duplicate task can be removed by doing the same trick you did above to determine whether to use `_xpack/license` or just `_license`.

File: templates/elasticsearch.yml.j2 [hunk 16]
```diff
@@ -11,6 +10,10 @@ cluster.name: elasticsearch
 node.name: {{inventory_hostname}}
 {% endif %}
 
+{% if (es_version is version_compare('7.0.0', '>=')) %}
+cluster.initial_master_nodes: {{inventory_hostname}}
```
Comment (Crazybus) [inline templates/elasticsearch.yml.j2]: I don't think this one should be here. We expect users to configure discovery themselves https://github.com/elastic/ansible-elasticsearch/#basic-elasticsearch-configuration

File: templates/elasticsearch.yml.j2 [hunk 17]
```diff
@@ -39,6 +42,10 @@ action.auto_create_index: {{ es_action_auto_create_index }}
 xpack.security.enabled: true
 {% endif %}
 
+{% if es_enable_xpack and es_api_basic_auth_username is defined and es_api_basic_auth_password is defined %}
+xpack.security.transport.ssl.enabled: true
```
Comment (Crazybus) [inline templates/elasticsearch.yml.j2]: Could you explain why this needs to be hardcoded in? The [default setting is `false`](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-settings.html#transport-tls-ssl-settings) which would be true for anybody who currently has this deployed which would make this a backwards incompatible change.

File: test/integration/helpers/serverspec/xpack_upgrade_spec.rb [hunk 18]
```diff
@@ -18,7 +18,7 @@
 
   describe 'security roles' do
     it 'should list the security roles' do
-      roles = curl_json('http://localhost:9200/_xpack/security/role', username='es_admin', password='changeMeAgain')
+      roles = curl_json("http://localhost:9200/_xpack/security/role", username='es_admin', password='changeMeAgain')
```
Comment (Crazybus) [inline test/integration/helpers/serverspec/xpack_upgrade_spec.rb]: Curious why these had to be switched to double quotes. The difference between single and double quotes in ruby is that it allows variable interpolation (which is not happening here). My best bet is that you originally were setting `es_api_scheme` here and removed it later.

File: tasks/xpack/security/elasticsearch-security-native.yml (hunk 10 above)
Comment (itsmed) [reply tasks/xpack/security/elasticsearch-security-native.yml]: Not sure if this means I should stop working so early/late when I'm not fully awake, or give in and not switch to vim 😄 either way, removed in https://github.com/elastic/ansible-elasticsearch/pull/479/commits/9b8aab8a4e1f6253ba36a4dcb3a48b8f96b339d2

File: tasks/elasticsearch-template.yml (hunk 8 above)
Comment (itsmed) [reply tasks/elasticsearch-template.yml]: Nope this was a `rebase` mistake, I think that https://github.com/elastic/ansible-elasticsearch/pull/479/commits/4fc4c6d0bef174a26d209770fbf79cfcad0d1f92 fixes it

File: tasks/elasticsearch-template.yml (hunk 9 above)
Comment (itsmed) [reply tasks/elasticsearch-template.yml]: Nope this was a `rebase` mistake, I think that https://github.com/elastic/ansible-elasticsearch/pull/479/commits/4fc4c6d0bef174a26d209770fbf79cfcad0d1f92 fixes it

File: README.md [hunk 19]
```diff
@@ -183,8 +183,8 @@ A more complex example:
 
 #### Important Note
 
-**The role uses es_api_host and es_api_port to communicate with the node for actions only achievable via http e.g. to install templates and to check the NODE IS ACTIVE.  These default to "localhost" and 9200 respectively.
-If the node is deployed to bind on either a different host or port, these must be changed.**
+**The role uses es_api_scheme, es_api_host, and es_api_port to communicate with the node for actions only achievable via http e.g. to install templates and to check the NODE IS ACTIVE.  These def
+ault to "http", "localhost", and 9200 respectively. If the node is deployed to bind on either a different host or port, these must be changed.**
```
Comment (jmlrt) [inline README.md]: I guess this line break was commited by mistake

**Comment by jmlrt** [comment]
> How does one setup their https certificate and node to node TLS

@Crazybus @itsmed 

We already have a PR waiting review which seem to implement TLS transport with certificate https://github.com/elastic/ansible-elasticsearch/pull/597

However just need to ensure than having that in different PR make sense.

**Comment by jmlrt** [comment]
I'm closing this one now that #620 is merged
